File: i_pylox/interpreter.py
from expr import ExprVisitor
from stmt import StmtVisitor
from token import TokenType
from env import Environment
from error import handleRuntimeError, RuntimeError
from callable import LoxCallable, LoxFunction, Return
from time import time


class ClockCallable(LoxCallable):

    def arity(self):
        return 0

    def call(self, interpreter, arguments):
        return time()

    def __str__(self):
        return '<native fn>'


class Interpreter(ExprVisitor, StmtVisitor):

    globals = Environment()
    environment = globals
    locals = {}

    def __init__(self):
        self.globals.define('clock', ClockCallable())

    def interpret(self, statements):
        try:
            for statement in statements:
                self.execute(statement)
        except RuntimeError as e:
            handleRuntimeError(e)

    def visitLiteralExpr(self, expr):
        return expr.value

    def visitLogicalExpr(self, expr):
        left = self.evaluate(expr.left)

        if expr.operator.type == TokenType.OR:
            if self.is_truthy(left): return left
        else:
            if not self.is_truthy(left): return left

        return self.evaluate(expr.right)

    def visitGroupingExpr(self, expr):
        return self.evaluate(expr.expression)

    def visitUnaryExpr(self, expr):
        right = self.evaluate(expr.right)

        match expr.operator.type:
            case TokenType.BANG: return not self.is_truthy(right)
            case TokenType.MINUS:
                self.check_number_operand(expr.operator, right)
                return -right
            case _: return None

    def visitVariableExpr(self, expr):
        return self.lookup_variable(expr.name, expr)

    def lookup_variable(self, name, expr):
        distance = self.locals.get(expr)
        if distance is not None:
            return self.environment.get_at(distance, name.lexeme)
        else:
            return self.globals.get(name)

    def check_number_operand(self, operator, operand):
        if isinstance(operand, float): return
        raise RuntimeError(operator, "Operand must be a number.")

    def check_number_operands(self, operator, left, right):
        if isinstance(left, float) and isinstance(right, float): return
        raise RuntimeError(operator, "Operands must be numbers.")

    def is_truthy(self, object):
        if object is None: return False
        if isinstance(object, bool): return object
        return True

    def is_equal(self, a, b):
        if a is None and b is None: return True
        if a is None: return False

        return a == b

    def stringify(self, object):
        if object is None: return 'nil'

        if isinstance(object, float):
            text = str(object)
            if text.endswith('.0'):
                text = text[0:-2]

            return text

        if isinstance(object, bool):
            return str(object).lower()

        return str(object)

    def evaluate(self, expr):
        return expr.accept(self)

    def execute(self, stmt):
        stmt.accept(self)

    def resolve(self, expr, depth):
        self.locals[expr] = depth

    def execute_block(self, statements, environment):
        previous = self.environment
        try:
            self.environment = environment

            for statement in statements:
                self.execute(statement)
        finally:
            self.environment = previous

    def visitBlockStmt(self, stmt):
        self.execute_block(stmt.statements, Environment(self.environment))

    def visitExpressionStmt(self, stmt):
        self.evaluate(stmt.expression)

    def visitFunctionStmt(self, stmt):
        function = LoxFunction(stmt, self.environment)
        self.environment.define(stmt.name.lexeme, function)
        return None

    def visitIfStmt(self, stmt):
        if self.is_truthy(self.evaluate(stmt.condition)):
            self.execute(stmt.then_branch)
        elif stmt.else_branch is not None:
            self.execute(stmt.else_branch)
        return None

    def visitPrintStmt(self, stmt):
        value = self.evaluate(stmt.expression)
        print(self.stringify(value))

    def visitReturnStmt(self, stmt):
        value = None
        if stmt.value is not None:
            value = self.evaluate(stmt.value)

        raise Return(value)

    def visitVarStmt(self, stmt):
        value = None
        if stmt.initializer is not None:
            value = self.evaluate(stmt.initializer)

        self.environment.define(stmt.name.lexeme, value)

    def visitWhileStmt(self, stmt):
        while self.is_truthy(self.evaluate(stmt.condition)):
            self.execute(stmt.body)

        return None

    def visitAssignExpr(self, expr):
        value = self.evaluate(expr.value)

        distance = self.locals.get(expr)
        if distance is not None:
            self.environment.assign_at(distance, expr.name, value)
        else:
            self.globals.assign(expr.name, value)

        return value

    def visitBinaryExpr(self, expr):
        left = self.evaluate(expr.left)
        right = self.evaluate(expr.right)

        match expr.operator.type:
            case TokenType.GREATER:
                self.check_number_operands(expr.operator, left, right)
                return left > right
            case TokenType.GREATER_EQUAL:
                self.check_number_operands(expr.operator, left, right)
                return left >= right
            case TokenType.LESS:
                self.check_number_operands(expr.operator, left, right)
                return left < right
            case TokenType.LESS_EQUAL:
                self.check_number_operands(expr.operator, left, right)
                return left <= right
            case TokenType.BANG_EQUAL: return not self.is_equal(left, right)
            case TokenType.EQUAL_EQUAL: return self.is_equal(left, right)
            case TokenType.MINUS:
                self.check_number_operands(expr.operator, left, right)
                return left - right
            case TokenType.PLUS:
                if isinstance(left, float) and isinstance(right, float):
                    return left + right

                if isinstance(left, str) and isinstance(right, str):
                    return left + right

                raise RuntimeError(expr.operator,
                    "Operands must be two numbers or two strings.")
            case TokenType.SLASH:
                self.check_number_operands(expr.operator, left, right)
                return left / right
            case TokenType.STAR:
                self.check_number_operands(expr.operator, left, right)
                return left * right

        return None

    def visitCallExpr(self, expr):
        callee = self.evaluate(expr.callee)

        arguments = []
        for argument in expr.arguments:
            arguments.append(self.evaluate(argument))

        if not isinstance(callee, LoxCallable):
            raise RuntimeError(expr.paren, "Can only call functions and classes.")

        if len(arguments) != callee.arity():
            raise RuntimeError(expr.paren,
                "Expected {} arguments but got {}.".format(callee.arity(), len(arguments)))

        return callee.call(self, arguments)

File: jlox/src/main/java/ahodanenok/craftinginterpeters/lox/Interpreter.java
package ahodanenok.craftinginterpreters.lox;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.LinkedList;
import java.util.List;
import java.util.Map;
import java.util.Objects;

class Interpreter implements Expression.Visitor<Object>, Statement.Visitor<Void> {

    Environment globals = new Environment();
    private Environment environment = globals;
    private Map<Expression, Integer> locals = new HashMap<>();
    private LinkedList<Boolean> loopBroken = new LinkedList<>();

    Interpreter() {
        globals.define("clock", new LoxCallable() {
            @Override
            public int arity() {
                return 0;
            }

            @Override
            public Object call(Interpreter interpreter, List<Object> arguments) {
                return System.currentTimeMillis() / 1000.0;
            }

            @Override
            public String toString() {
                return "<native fn>";
            }
        });
    }

    void interpret(List<Statement> program) {
        try {
            for (Statement statement : program) {
                execute(statement);
            }
        } catch (RuntimeError e) {
            Lox.runtimeError(e.token, e.getMessage());
        }
    }

    void resolve(Expression expression, int distance) {
        locals.put(expression, distance);
    }

    @Override
    public Void visitExprStatement(Statement.Expr statement) {
        evaluate(statement.expression);
        return null;
    }

    @Override
    public Void visitPrintStatement(Statement.Print statement) {
        Object value = evaluate(statement.expression);
        System.out.println(stringify(value));
        return null;
    }

    @Override
    public Void visitVarStatement(Statement.Var statement) {
        Object value = null;
        if (statement.initializer != null) {
            value = evaluate(statement.initializer);
        }
        environment.define(statement.name.lexeme, value);
        if (statement.initializer != null) {
            environment.markInitialized(statement.name.lexeme);
        }

        return null;
    }

    @Override
    public Void visitBlockStatement(Statement.Block statement) {
        executeBlock(statement.statements, new Environment(environment));
        return null;
    }

    void executeBlock(List<Statement> statements, Environment currentEnvironment) {
        Environment previousEnvironment = this.environment;
        try {
            this.environment = currentEnvironment;
            for (Statement statement : statements) {
                execute(statement);
            }
        } finally {
            this.environment = previousEnvironment;
        }
    }

    @Override
    public Void visitIfStatement(Statement.If statement) {
        if (isTruthy(evaluate(statement.condition))) {
            execute(statement.thenBranch);
        } else if (statement.elseBranch != null) {
            execute(statement.elseBranch);
        }
        return null;
    }

    @Override
    public Void visitWhileStatement(Statement.While statement) {
        loopBroken.push(false);
        try {
            while (isTruthy(evaluate(statement.condition)) && !loopBroken.peek()) {
                execute(statement.body);
            }
        } finally {
            loopBroken.pop();
        }
        return null;
    }

    @Override
    public Void visitBreakStatement(Statement.Break statement) {
        if (loopBroken.size() == 0) {
            throw new RuntimeError(statement.keyword, "No enclosing loop.");
        }
        loopBroken.set(0, true);
        return null;
    }

    @Override
    public Void visitFunctionStatement(Statement.Function statement) {
        LoxFunction function = new LoxFunction(statement, environment, false);
        environment.define(statement.name.lexeme, function);
        return null;
    }

    @Override
    public Void visitReturnStatement(Statement.Return statement) {
        Object value = null;
        if (statement.expression != null) {
            value = evaluate(statement.expression);
        }

        throw new Return(value);
    }

    @Override
    public Void visitClassStatement(Statement.Class statement) {
        Object parent = null;
        if (statement.parent != null) {
            parent = evaluate(statement.parent);
            if (!(parent instanceof LoxClass)) {
                throw new RuntimeError(statement.parent.name,
                    "Superclass must be a class.");
            }
        }

        environment.define(statement.name.lexeme, null);
        if (statement.parent != null) {
            environment = new Environment(environment);
            environment.define("super", parent);
        }

        Map<String, LoxFunction> methods = new HashMap<>();
        for (Statement.Function method : statement.methods) {
            methods.put(method.name.lexeme,
                new LoxFunction(method, environment, method.name.lexeme.equals("init")));
        }

        if (statement.parent != null) {
            environment = environment.parent;
        }

        LoxClass klass = new LoxClass(statement.name.lexeme, (LoxClass) parent, methods);
        environment.assign(statement.name, klass);
        return null;
    }

    @Override
    public Object visitLiteralExpression(Expression.Literal expression) {
        return expression.value;
    }

    @Override
    public Object visitUnaryExpression(Expression.Unary expression) {
        Object value = evaluate(expression.expression);
        return switch (expression.operator.type) {
            case MINUS -> {
                checkNumberOperand(expression.operator, value);
                yield -(double) value;
            }
            case BANG -> !isTruthy(value);
            default -> null;
        };
    }

    @Override
    public Object visitBinaryExpression(Expression.Binary expression) {
        Object left = evaluate(expression.left);
        Object right = evaluate(expression.right);
        return switch (expression.operator.type) {
            case GREATER_EQUAL -> {
                checkNumberOperands(expression.operator, left, right);
                yield (double) left >= (double) right;
            }
            case GREATER -> {
                checkNumberOperands(expression.operator, left, right);
                yield (double) left > (double) right;
            }
            case LESS_EQUAL -> {
                checkNumberOperands(expression.operator, left, right);
                yield (double) left <= (double) right;
            }
            case LESS -> {
                checkNumberOperands(expression.operator, left, right);
                yield (double) left < (double) right;
            }
            case BANG_EQUAL -> !isEqual(left, right);
            case EQUAL_EQUAL -> isEqual(left, right);
            case MINUS -> {
                checkNumberOperands(expression.operator, left, right);
                yield (double) left - (double) right;
            }
            case PLUS -> {
                if (left instanceof Double a && right instanceof Double b) {
                    yield a + b;
                }

                if (left instanceof String a && right instanceof String b) {
                    yield a + b;
                }
                // conflicts with the test suite
                // else if (left instanceof String || right instanceof String) {
                //     yield stringify(left) + stringify(right);
                // }

                throw new RuntimeError(
                    expression.operator,
                    "Operands must be two numbers or two strings.");
            }
            case STAR -> {
                checkNumberOperands(expression.operator, left, right);
                yield (double) left * (double) right;
            }
            case SLASH -> {
                checkNumberOperands(expression.operator, left, right);
                yield (double) left / (double) right;
            }
            default -> null;
        };
    }

    @Override
    public Object visitTernaryExpression(Expression.Ternary expression) {
        Object value = evaluate(expression.condition);
        if (isTruthy(value)) {
            return evaluate(expression.left);
        } else {
            return evaluate(expression.right);
        }
    }

    @Override
    public Object visitGroupingExpression(Expression.Grouping expression) {
        return evaluate(expression.expression);
    }

    @Override
    public Object visitVariableExpression(Expression.Variable expression) {
        return lookupVariable(expression.name, expression);
    }

    private Object lookupVariable(Token name, Expression expression) {
        Integer distance = locals.get(expression);
        if (distance != null) {
            return environment.getAt(distance, name.lexeme);
        } else {
            return globals.get(name);
        }
    }

    @Override
    public Object visitAssignExpression(Expression.Assign expression) {
        Object value = evaluate(expression.expression);
        Integer distance = locals.get(expression);
        if (distance != null) {
            environment.assignAt(distance, expression.name, value);
        } else {
            globals.assign(expression.name, value);
        }
        environment.markInitialized(expression.name.lexeme);
        return value;
    }

    @Override
    public Object visitLogicalExpression(Expression.Logical expression) {
        Object value = evaluate(expression.left);
        if (expression.operator.type == TokenType.OR) {
            if (isTruthy(value)) {
                return value;
            }
        } else if (expression.operator.type == TokenType.AND) {
            if (!isTruthy(value)) {
                return value;
            }
        } else {
            throw new RuntimeError(expression.operator, "Expect logical operator.");
        }

        return evaluate(expression.right);
    }

    @Override
    public Object visitCallExpression(Expression.Call expression) {
        Object callee = evaluate(expression.callee);
        List<Object> arguments = new ArrayList<>();
        for (Expression argument : expression.arguments) {
            arguments.add(evaluate(argument));
        }

        if (callee instanceof LoxCallable callable) {
            if (callable.arity() != arguments.size()) {
                throw new RuntimeError(expression.paren,
                    String.format(
                        "Expected %d arguments but got %d.",
                        callable.arity(),
                        arguments.size()));
            }

            return callable.call(this, arguments);
        }

        throw new RuntimeError(expression.paren,
            "Can only call functions and classes.");
    }

    @Override
    public Object visitLambdaExpression(Expression.Lambda expression) {
        return new LoxLambda(expression, environment);
    }

    @Override
    public Object visitGetExpression(Expression.Get expression) {
        Object object = evaluate(expression.object);
        if (object instanceof LoxInstance instance) {
            return instance.get(expression.name);
        }

        throw new RuntimeError(expression.name,
            "Only instances have properties.");
    }

    @Override
    public Object visitSetExpression(Expression.Set expression) {
        Object object = evaluate(expression.object);
        if (object instanceof LoxInstance instance) {
            Object value = evaluate(expression.value);
            instance.set(expression.name, value);
            return value;
        }

        throw new RuntimeError(expression.name, "Only instances have fields.");
    }

    @Override
    public Object visitThisExpression(Expression.This expression) {
        return lookupVariable(expression.keyword, expression);
    }

    @Override
    public Object visitSuperExpression(Expression.Super expression) {
        int distance = locals.get(expression);
        LoxClass parent = (LoxClass) environment.getAt(distance, "super");
        LoxInstance object =
            (LoxInstance) environment.getAt(distance - 1, "this");
        LoxFunction method = parent.findMethod(expression.method.lexeme);
        if (method == null) {
            throw new RuntimeError(expression.method,
                "Undefined property '" + expression.method.lexeme + "'.");
        }

        return method.bind(object);
    }

    private Object evaluate(Expression expression) {
        return expression.accept(this);
    }

    private void execute(Statement statement) {
        if (loopBroken.size() > 0 && loopBroken.peek()) {
            return;
        }

        statement.accept(this);
    }

    private boolean isTruthy(Object value) {
        if (value == null) {
            return false;
        } else if (value instanceof Boolean) {
            return (boolean) value;
        } else {
            return true;
        }
    }

    private boolean isEqual(Object a, Object b) {
        return Objects.equals(a, b);
    }

    private void checkNumberOperand(Token operator, Object operand) {
        if (!(operand instanceof Double)) {
            throw new RuntimeError(operator, "Operand must be a number.");
        }
    }

    private void checkNumberOperands(Token operator, Object left, Object right) {
        if (!(left instanceof Double) || !(right instanceof Double)) {
            throw new RuntimeError(operator, "Operands must be numbers.");
        }
    }

    private String stringify(Object value) {
        if (value == null) {
            return "nil";
        }

        if (value instanceof Double n) {
            String text = value.toString();
            if (text.endsWith(".0")) {
                return text.substring(0, text.length() - 2);
            } else {
                return text;
            }
        }

        return value.toString();
    }
}
